### PR TITLE
Change response body type to Cow<'a, [u8]>

### LINKED ISSUE #72: Can't use owned bytes for body in Server handler?
It seems like you can't use owned data (e.g., `String` or `Vec<u8>`) for the response body in the `Server` handler. 

Unless I am missing something, this seems like a huge limitation: it prevents you from doing anything with the request data (at least easily).

It seems like you could use `Cow<'static, [u8]>` for the response body, or `Bytes`.

Am I missing something?

### PR DESCRIPTION
**Pull request by kardeiz**
Changed the response body type to `Cow<'a, [u8]>`, to resolve #72.

Allows one to create response body data within the handler, which wasn't possible before.

Doesn't add much complexity or any dependencies, but is a breaking change, so feel free to ignore/close this PR!

### CODE REVIEW

File: examples/server.rs [hunk 1]
```diff
@@ -13,8 +13,8 @@ fn main() {
     let port = "7878";
 
     let server = Server::new(|request, mut response| {
-        info!("Request received. {} {}", request.method(), request.uri());
-        Ok(response.body("Hello Rust!".as_bytes())?)
+        let msg = format!("The path you requested was: '{}'", request.uri().path());
```
Comment (steveklabnik) [inline examples/server.rs]: this changes the behavior of the example; I'd prefer to keep it the same.

**Comment by steveklabnik** [comment]
The tests are still failing. 

Gotta talk this over with @ashleygwilliams , but I can see why we'd want this. I wonder if we shouldn't go full generic, something like

```rust
struct Server<'a, T: Into<Cow<'a, [u8]> + 'a> {
    handler: fn(Request<&[u8]>, ResponseBuilder) -> Result<Response<T>, Error>,
} 
```

and then move the `into()` into where it's invoked. A bit of complexity on our side for an easier api on their side.

**Comment by kardeiz** [comment]
Is that even possible? The response body will have to be a concrete type, right?

What would be nice for this case would be for `http::response::Builder::body` to take a `body: B` where `B: Into<T>`, and then call `Into::into` on it before it when it returns the response.

File: examples/server.rs (hunk 1 above)
Comment (kardeiz) [reply examples/server.rs]: sorry, forgot I had changed that when I was doing a quick check of this functionality

Comment (ashleygwilliams) [reply examples/server.rs]: agree. the example staying the same is like a test passing, IMO.

**Comment by ashleygwilliams** [comment]
i haven't had too much time to read and think on this but one thing i want to remind everyone is that the user interface should be kept as close to what it is now, and ideally simple. i'll take a closer look at this but i tend to agree with @steveklabnik's points.

File: examples/server.rs [hunk 2]
```diff
@@ -14,7 +14,7 @@ fn main() {
 
     let server = Server::new(|request, mut response| {
         info!("Request received. {} {}", request.method(), request.uri());
-        Ok(response.body("Hello Rust!".as_bytes())?)
+        Ok(response.body("Hello Rust!".as_bytes().into())?)
```
Comment (ashleygwilliams) [inline examples/server.rs]: i'm not a fan of this `into()`. this is why i do not like this approach.

Comment (ashleygwilliams) [reply examples/server.rs]: (sorry if that sounded gruff, lots of activity on this right now and i really want to defend the simple interface. i appreciate your attention and care in this PR! it just might not be the direction we want to go in at the moment)

Comment (kardeiz) [reply examples/server.rs]: that's okay, i understand your concern about keeping things simple. i would like to be able to use this library, but not being able to generate a response body from the handler would be a blocker for just about any project i would use this for.

**Comment by kardeiz** [comment]
I was wrong, it is possible to do this "fully" generic as @steveklabnik suggested; see the latest commits.

File: src/lib.rs [hunk 3]
```diff
@@ -51,12 +53,12 @@ pub use error::Error;
 ///
 /// This is the core type of this crate, and is used to create a new
 /// server and listen for connections.
-pub struct Server {
-    handler: fn(Request<&[u8]>, ResponseBuilder) -> Result<Response<&[u8]>, Error>,
+pub struct Server<T> {
```
Comment (steveklabnik) [inline src/lib.rs]: i prefer to repeat the bounds on the struct as well; as this means you could construct a `Server` that doesn't have the `listen` method!

Comment (kardeiz) [reply src/lib.rs]: Doing so means `Server` will have to have a `PhantomData` field. Is it worth it?

Also is it fine to have a `marker: PhantomData<&'a ()>` here, or does it have to be `marker: PhantomData<&'a T>`? Both compile just fine, but the latter requires adding the `Sync` trait to `T`.

Comment (steveklabnik) [reply src/lib.rs]: hm, does it? That strikes me as odd....

Comment (steveklabnik) [reply src/lib.rs]: okay, i guess that we shouldn't do this, then

would you mind adding an example in the examples directory that returns owned stuff? After that, this is good to go IMO

**Comment by kardeiz** [comment]
I've added an example, `dynamic_response`, in the `examples` folder.

**Comment by steveklabnik** [comment]
Hey @kardeiz ! Thanks so much for this PR, and sorry for the long review time; it's been a busy week for us, with two conferences!

I've merged this in manually as https://github.com/steveklabnik/simple-server/commit/7bfd93840bf695b8f55992154df297a07b669297; I wanted to squash stuff down. We'll be cutting a new release with this, and then using it to teach a RustBridge tomorrow; your work will help a bunch of new people learn Rust! 🎊 ❤️

**Pull request closed without merging**